exonum-java-binding/testkit/src/test/java/com/exonum/binding/testkit/TestKitExtension.java

```diff
@@ -61,11 +58,87 @@ public boolean supportsParameter(ParameterContext parameterContext,
   public Object resolveParameter(
       ParameterContext parameterContext, ExtensionContext extensionContext)
       throws ParameterResolutionException {
-    return getStore(extensionContext).get(KEY, TestKit.class);
+    checkExtensionContext(extensionContext);
+    CloseableTestKit closeableTestKit =
+        getStore(extensionContext).get(KEY, CloseableTestKit.class);
+    // Check if any TestKit configuration annotations are used
+    boolean annotationsUsed = annotationsUsed(parameterContext);
+    TestKit testKit;
+    if (closeableTestKit == null) {
+      if (annotationsUsed) {
+        updateTestKitBuilder(parameterContext, extensionContext);
+      }
+      testKit = testKitBuilder.build();
+      CloseableTestKit newCloseableTestKit = new CloseableTestKit(testKit);
+      getStore(extensionContext).put(KEY, newCloseableTestKit);
+    } else {
+      // Throw an exception if TestKit was already instantiated in this context, but user tries to
+      // reconfigure it
+      if (annotationsUsed) {
+        throw new RuntimeException("TestKit was parameterized with annotations after being"
+            + " instantiated in " + extensionContext.getDisplayName());
+      }
+      testKit = closeableTestKit.getTestKit();
+    }
+    return testKit;
+  }
+
+  /**
+   * Check the extension context and throw if TestKit is injected in @BeforeAll or @AfterAll.
+   */
+  private void checkExtensionContext(ExtensionContext extensionContext) {
+    Optional<Method> testMethod = extensionContext.getTestMethod();
+    testMethod.orElseThrow(() ->
+        new RuntimeException("TestKit can't be injected in @BeforeAll or @AfterAll"));
```

Comment:
On the question of whether we shall or shall not forbid it:

"If you do need the same Testkit instance for all tests in the class, just use the builder directly:

private static final Testkit testkit = Testkit.newBuilder()
    .withService(YourServiceModule.class)
    .build();
"

… but as I wrote this instruction I remembered that the user will also need to write `@AfterAll` that *destroys* the testkit! So maybe we shall support injection in `BeforeAll` after all, shan't we?

Reply:
Yes, but isn't the idea of `@Test` cases is that they are independent? Why would a user want to use the same TestKit? It's usually used to create some blocks and then review the state of the system, so using the same TestKit would lead to errors.

But we could document that and it would be easier for us - no need to throw an exception in `@Before/AfterAll` injections.

Reply:
So the questions are:
1. Could anyone (@bullet-tooth) think of a realistic use-case when one needs a shared testkit for all methods?
2. Shall we restrict such usage (as error-prone) if we can't think of such use-cases?

I can think of a scenario when setup is performed in `@BeforeAll` and all tests perform only read-only operations on testkit — but how realistic that is, and how resistant to future additions of new tests performing write operations? Also, this scenario would also work with `@BeforeEach`, somewhat slower, but way more reliable.

Reply:
Also I'd prefer to explicitly prohibit injecting in `@AfterAll`, as it doesn't really make sense (it'll just create one more TestKit if injected). Although I agree that users might want to perform only read-only operations in different tests, it seems like a very error-prone approach. I think `@BeforeEach` is more reliable indeed even if slow.

Reply:
> it'll just create one more TestKit if injected

It won't — contexts are nested, everything in the parent context (`@BeforeAll`) is accessible in children (`@Beach`, `@Test`).

Reply:
Ok, let's forbid but let's put something along the lines of

```
Testkit cannot be injected in @BeforeAll/AfterAll because it is a stateful, mutable, object and sharing it between all tests is error-prone. Consider injecting it in @BeforeEach instead.

If you *do* need the same instance for all tests — just use `Testkit#builder` directly. Don't forget to destroy it in @AfterEach.
```

Reply:
> It won't — contexts are nested, everything in the parent context (@BeforeAll) is accessible in children (@Beach, @Test).

I was talking about `@AfterAll`. If we inject in `@BeforeAll`, we'd have same TestKit in all tests - but if we inject in `@AfterAll`, we won't - every Test will have its own instance, including `@AfterAll` method.

Reply:
I see, in this case — yes, but it makes no apparent sense to use that :-)